pkg/metrics/roundtripper.go

```diff
@@ -0,0 +1,151 @@
+package metrics
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/operator-framework/operator-marketplace/pkg/defaults"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// appRegistryRoundTripperCounter wraps a RoundTripperFunc with functions that record prometheus metrics
+// when reaching out to an AppRegistry.
+func appRegistryRoundTripperCounter(counter *prometheus.CounterVec, next http.RoundTripper) promhttp.RoundTripperFunc {
+	return promhttp.RoundTripperFunc(func(r *http.Request) (*http.Response, error) {
+		resp, err := next.RoundTrip(r)
+		if err == nil {
+			if quayNamespace, ok := r.URL.Query()["namespace"]; ok && len(quayNamespace) == 1 {
+				counter.With(labels(r.Method, resp.StatusCode, quayNamespace[0])).Inc()
+			}
+		}
+		return resp, err
+	})
+}
+
+// labels adds the expected labels for the AppRegistry Counter.
+func labels(reqMethod string, status int, quayNamespace string) prometheus.Labels {
+	labels := prometheus.Labels{}
+	labels["code"] = sanitizeCode(status)
+	labels["method"] = strings.ToLower(reqMethod)
+	labels["opsrc"] = getOpsrcLabelValue(quayNamespace)
+
+	return labels
+}
+
+// getOpsrcLabelValue returns the quayNamespace value if it
+// matches a default OperatorSource. Otherwise, a generic
+// name is returned to prevent leaking customer data to telemetry.
+func getOpsrcLabelValue(quayNamespace string) string {
+	if defaults.IsDefaultSource(quayNamespace) {
+		return quayNamespace
+	}
+	return "non-default-opsrc"
+}
+
+// If the wrapped http.Handler has not set a status code, i.e. the value is
+// currently 0, santizeCode will return 200, for consistency with behavior in
+// the stdlib.
+//
+// Taken from the prometheus operator project: https://github.com/prometheus/client_golang/blob/b4781a8cbdccaef5a0a785bea59ca3d5adc1554f/prometheus/promhttp/instrument_server.go#L348-L447
+func sanitizeCode(s int) string {
```

Comment:
Do we really need all of these case statements besides the special 0 case?

Reply:
That's a good question - I had just taken the same approach as the prometheus-operator code but would be happy to just call ` strconv.Itoa()`.

Reply:
I see your point now, I will update the switch to:
```
switch s {
	case 200, 0:
		return "200"
	default:
		return strconv.Itoa(s)
	}
}
```

Reply:
If the current value is `0` we do want to set it to `200` before we call `strconv.Itoa()`, right?